### DISCUSSION TITLE
How to make a combined action run on either PR or Push event?

### DISCUSSION THREAD

**Discussion by iamfotx** [community/community#26786] [Actions]
<p>I need a workflow, where I want action to run on each PR against master. And I also want to run that same action when a push (direct commit not PR merge commit) happens in master.</p>
<p>Current configurations looks like this:<br>
<img src="https://user-images.githubusercontent.com/20137243/181099895-727ea042-2bed-47a4-85c2-375a74232901.png" alt="image" data-base62-sha1="6STDvBssp8HJ4uzBeiAo3Ll0lX0" width="229" height="148"></p>
<p>The Problem:<br>
It makes those actions run twice i.e</p>
<ol>
<li>When PR is being created against master (that makes sense as I’m listening to on PR event)</li>
<li>When that PR gets merged into master (I don’t want to re run the action this time as It’s already in passing state from PR)</li>
</ol>
<p>Expected Behavior:<br>
I’m expecting action to not run twice, I want to run the action like this:</p>
<ol>
<li>When a PR gets opened against master, run the action.</li>
<li>When a direct commit happens to master (again not on the PR merge commit as it will make the action to run twice).</li>
</ol>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/iamfotx">@iamfotx</a>,</p>
<p>When a commit is pushed from a merged PR, the automatically generated the head commit message generally follows the format “<strong>Merge pull request #&lt;pr_number&gt; from /&lt;source_branch&gt;\n\n&lt;pr_title&gt;</strong>”.<br>
For example:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/3/347c5a26e7bcda21e035bccd68001d0bd2cbba86.png" data-download-href="/uploads/short-url/7ujfBE6HTCHGsEU7cdISgNTUAL4.png?dl=1" title="head_commit_message" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181099898-1c0d98fd-9715-47fe-b855-f12f196bf3d2.png" alt="head_commit_message" data-base62-sha1="7ujfBE6HTCHGsEU7cdISgNTUAL4" width="690" height="82" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/347c5a26e7bcda21e035bccd68001d0bd2cbba86_2_690x82.png, https://aws1.discourse-cdn.com/github/original/2X/3/347c5a26e7bcda21e035bccd68001d0bd2cbba86.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/3/347c5a26e7bcda21e035bccd68001d0bd2cbba86.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/347c5a26e7bcda21e035bccd68001d0bd2cbba86_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">head_commit_message</span><span class="informations">998×119 10.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>You can use the expression “<strong>github.event.head_commit.message</strong>” to get the head commit message from the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener">github context</a>.<br>
You can use the following <strong>if</strong> conditional to skip all the jobs in the workflow when the push is from the PR merged.</p>
<pre><code class="lang-yaml">if: ${{ github.event_name == 'pull_request' || (github.event_name == 'push' &amp;&amp; !contains( github.event.head_commit.message, 'Merge pull request')) }}
</code></pre>
<p>A simple demo:</p>
<pre><code class="lang-yaml">on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master

jobs:
  job1:
    if: ${{ github.event_name == 'pull_request' || (github.event_name == 'push' &amp;&amp; !contains( github.event.head_commit.message, 'Merge pull request')) }}
    . . .

  job2:
    if: ${{ github.event_name == 'pull_request' || (github.event_name == 'push' &amp;&amp; !contains( github.event.head_commit.message, 'Merge pull request')) }}
    . . .

  jobN:
    if: ${{ github.event_name == 'pull_request' || (github.event_name == 'push' &amp;&amp; !contains( github.event.head_commit.message, 'Merge pull request')) }}
    . . .
</code></pre>

**Comment by iamfotx** [comment]
<p>Thank you so much for the reply! It worked like charm.</p>
<p>But there’s a problem with git checks:</p>
<p>This is how my commit looks in an Active PR…<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/fa7103673674537e5f2d6ea74455bed6f1bf1998.png" data-download-href="/uploads/short-url/zJvuQiKxZZ1dtQEZkWpEImt0lKg.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/20137243/181099899-903a3216-a948-4334-bbf8-1023d1e783df.png" alt="image" data-base62-sha1="zJvuQiKxZZ1dtQEZkWpEImt0lKg" width="690" height="21" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/f/fa7103673674537e5f2d6ea74455bed6f1bf1998_2_690x21.png, https://aws1.discourse-cdn.com/github/original/2X/f/fa7103673674537e5f2d6ea74455bed6f1bf1998.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/f/fa7103673674537e5f2d6ea74455bed6f1bf1998.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/fa7103673674537e5f2d6ea74455bed6f1bf1998_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">856×27 3.87 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>When I merge this PR into master:<br>
<img src="https://user-images.githubusercontent.com/20137243/181099902-ce6d5c46-0384-4b91-9b1c-9e73eef72f79.png" alt="image" data-base62-sha1="6xr8wWnrNOl0GcvBY4EyyW8SiY8" width="369" height="31"></p>
<p>I’ve lost this green status check mark. And I think when using circleci/codeship, we didn’t to re-run the workflow on the master branch again rather it picks the status of the PR.</p>

**Comment by airtower-luna** [comment]
<p>The second screenshot doesn’t have a commit ID, but I assume it’s showing the merge commit? If so, that’s exactly why there is no check mark: The workflow didn’t run for the merge commit. When you look at the commit history the original commit should still have the check mark.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/iamfotx">@iamfotx</a>,</p>
<p>The green hook icon (:ballot_box_with_check:) under a commit means all the checks (include jobs in the workflows) run on this commit ref have passed.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/6/66b6736088a9a6f79e8ea3d55d18cfa328ff3aa2.png" data-download-href="/uploads/short-url/eEDBfr5DipJuRg8ruJfudRqGSIO.png?dl=1" title="checks_passed" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181099904-4ad52d19-3f35-4d55-bb3d-19d26de8d3ff.png" alt="checks_passed" data-base62-sha1="eEDBfr5DipJuRg8ruJfudRqGSIO" width="690" height="184" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/6/66b6736088a9a6f79e8ea3d55d18cfa328ff3aa2_2_690x184.png, https://aws1.discourse-cdn.com/github/original/2X/6/66b6736088a9a6f79e8ea3d55d18cfa328ff3aa2.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/6/66b6736088a9a6f79e8ea3d55d18cfa328ff3aa2.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/6/66b6736088a9a6f79e8ea3d55d18cfa328ff3aa2_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">checks_passed</span><span class="informations">744×199 20.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>If some checks were not successful, a red cross icon (:negative_squared_cross_mark:) is displayed.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/b/b3e387e29caafb32a49c3d4a0422fb8f4675e4f5.png" data-download-href="/uploads/short-url/pFmWSFaV2mPYhnJjKGe7EAWXkEJ.png?dl=1" title="checks_failed" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181099905-80089ea6-cf67-4ee3-8374-cf2edc461512.png" alt="checks_failed" data-base62-sha1="pFmWSFaV2mPYhnJjKGe7EAWXkEJ" width="690" height="192" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/b/b3e387e29caafb32a49c3d4a0422fb8f4675e4f5_2_690x192.png, https://aws1.discourse-cdn.com/github/original/2X/b/b3e387e29caafb32a49c3d4a0422fb8f4675e4f5.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/b/b3e387e29caafb32a49c3d4a0422fb8f4675e4f5.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/b3e387e29caafb32a49c3d4a0422fb8f4675e4f5_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">checks_failed</span><span class="informations">728×203 24.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>As <a class="mention" href="/airtower-luna">@airtower-luna</a> mentioned, because you have skipped all the checks, there is not checks status displayed under the commit.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/iamfotx">@iamfotx</a>,</p>
<p>How are things going?<br>
Are our above explanations helpful to you to understand?<br>
If you have any questions about this topic, feel free to tell us.</p>

**Comment by iamfotx** [comment]
<p>Ty!, things are in pretty good shape.</p>
